### PR TITLE
bpo-9938: Add optional keyword argument exit_on_error to argparse.ArgumentParser

### PR DESCRIPTION
**Pull request by shihai1991**
Co-Authored-by: Xuanji Li <xuanji@gmail.com>

https://bugs.python.org/issue9938

Automerge-Triggered-By: @matrixise

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2019-08-21-16-38-56.bpo-9938.t3G7N9.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Add ``exit_on_error`` in :class:`ArgumentParser`.
```
Comment (aeros) [inline Misc/NEWS.d/next/Library/2019-08-21-16-38-56.bpo-9938.t3G7N9.rst]: 3) Change the Misc/NEWS entry to specify that ``allow_abbrev`` is an optional keyword argument.

```suggestion
Add optional keyword argument ``exit_on_error`` for :class:`ArgumentParser`.
```

**Review by aeros — CHANGES_REQUESTED** [review]
Thanks for the PR @shihai1991! I could definitely see the addition of an ``exit_on_error`` argument being quite useful to give users more customization. 

I would recommend the following additions:

1) Update the docstring by adding the following on line 1626 (in ``Lib/argparse.py``):
```
- exit_on_error -- Determines whether or not argparser exits when an error occurs.
```

2) Update the docs for [ArgumentParser](https://docs.python.org/3.5/library/argparse.html#argparse.ArgumentParser).

**Comment by aeros167** [comment]
Also, I would recommend adjusting the PR title to ``bpo-9938: Add optional kwarg to ArgParser``. This will allow core devs to more easily identify what the PR is addressing.

**Comment by aeros167** [comment]
The Travis CI failure can be fixed by locally running ``make patchcheck`` to address the whitespace issue. The problem seems to be occurring from ``Lib/test/test_argparse.py``.

**Comment by shihai1991** [comment]
> Thanks for the PR @shihai1991! I could definitely see the addition of an `exit_on_error` argument being quite useful to give users more customization.
> 
> I would recommend the following additions:
> 
> 1. Update the docstring by adding the following on line 1626 (in `Lib/argparse.py`):
> 
> ```
> - exit_on_error -- Determines whether or not argparser exits when an error occurs.
> ```
> 
> 1. Update the docs for [ArgumentParser](https://docs.python.org/3.5/library/argparse.html#argparse.ArgumentParser).

Looks `Determines whether or not argparser exits with error when an error occurs` would better ;).

> - exit_on_error -- Determines whether or not argparser exits when an error occurs.

np.

File: Misc/NEWS.d/next/Library/2019-08-21-16-38-56.bpo-9938.t3G7N9.rst (hunk 1 above)
Comment (shihai1991) [reply Misc/NEWS.d/next/Library/2019-08-21-16-38-56.bpo-9938.t3G7N9.rst]: your desc looks more exact, thanks.

**Comment by shihai1991** [comment]
> Also, I would recommend adjusting the PR title to `[bpo-9938](https://bugs.python.org/issue9938): Add optional kwarg to ArgParser`. This will allow core devs to more easily identify what the PR is addressing.

the desc looks good to me. But why the title need add the bpo link now? the format of title have been changed?

File: Lib/argparse.py [hunk 2]
```diff
@@ -1623,6 +1623,8 @@ class ArgumentParser(_AttributeHolder, _ActionsContainer):
         - conflict_handler -- String indicating how to handle conflicts
         - add_help -- Add a -h/-help option
         - allow_abbrev -- Allow long options to be abbreviated unambiguously
+        - exit_on_error -- Determines whether or not argparser exits with
```
Comment (aeros) [inline Lib/argparse.py]:
```suggestion
        - exit_on_error -- Determines whether or not argparser exits with an
```

File: Doc/library/argparse.rst [hunk 3]
```diff
@@ -179,13 +179,19 @@ ArgumentParser objects
    * allow_abbrev_ - Allows long options to be abbreviated if the
      abbreviation is unambiguous. (default: ``True``)
 
+   * exit_on_error_ - Determines whether or not argparser exits with error
```
Comment (aeros) [inline Doc/library/argparse.rst]:
```suggestion
   * exit_on_error_ - Determines whether or not argparser exits with an error
```

File: Doc/library/argparse.rst [hunk 4]
```diff
@@ -179,13 +179,19 @@ ArgumentParser objects
...
+     when an error occur. (default: ``True``)
```
Comment (aeros) [inline Doc/library/argparse.rst]:
```suggestion
     when an error occurs. (default: ``True``)
```

**Review by aeros — APPROVED** [review]
@shihai1991 
> the desc looks good to me. But why the title need add the bpo link now? the format of title have been changed?

I may have been a bit unclear, apologies. I just included the bpo part so that you wouldn't accidentally remove it when editing the title and I didn't intend to include the link. It looks good now. 

Thanks for making the recommended changes.

I have a couple of minor grammar suggestions, but once those are added, everything documentation-related looks good as far as I can tell. (:

**Comment by shihai1991** [comment]
> @shihai1991
> 
> > the desc looks good to me. But why the title need add the bpo link now? the format of title have been changed?
> 
> I may have been a bit unclear, apologies. I just included the bpo part so that you wouldn't accidentally remove it when editing the title and I didn't intend to include the link. It looks good now.
> 
> Thanks for making the recommended changes.
> 
> I have a couple of minor grammar suggestions, but once those are added, everything documentation-related looks good as far as I can tell. (:

thanks, good job :)

**Comment by shihai1991** [comment]
cc @rhettinger

File: Doc/library/argparse.rst [hunk 5]
```diff
@@ -647,6 +653,28 @@ the help options::
      +h, ++help  show this help message and exit
 
 
+exit_on_error
+^^^^^^^^^^^^^
+
+Normally, when you pass an wrong argument list to the :meth:`~ArgumentParser.parse_args`
```
Comment (aeros) [inline Doc/library/argparse.rst]: A basic grammar correction would be to just change "an" to "a", but I think that "invalid" makes more sense in this context than "wrong":

```suggestion
Normally, when you pass an invalid argument list to the :meth:`~ArgumentParser.parse_args`
```

File: Doc/library/argparse.rst [hunk 6]
```diff
@@ -647,6 +653,28 @@ the help options::
...
+method of an :class:`ArgumentParser`, it will exit with error info.
+
+If user would like catch error by themselves, they can enable this feature by setting
```
Comment (aeros) [inline Doc/library/argparse.rst]: Minor grammar fix and "by themselves" to "manually":

```suggestion
If the user would like catch errors manually, they can enable this feature by setting
```

**Review by aeros — APPROVED** [review]

File: Doc/library/argparse.rst [hunk 7]
```diff
@@ -179,13 +179,19 @@ ArgumentParser objects
    * allow_abbrev_ - Allows long options to be abbreviated if the
      abbreviation is unambiguous. (default: ``True``)
 
+   * exit_on_error_ - Determines whether or not argparser exits with error info
```
Comment (matrixise) [inline Doc/library/argparse.rst]: argparser -> ArgumentParser

File: Doc/library/argparse.rst [hunk 8]
```diff
@@ -647,6 +653,28 @@ the help options::
      +h, ++help  show this help message and exit
 
 
+exit_on_error
+^^^^^^^^^^^^^
+
+Normally, when you pass an invalid argument list to the :meth:`~ArgumentParser.parse_args`
+method of an :class:`ArgumentParser`, it will exit with error info.
+
+If the user would like catch errors manually, they can enable this feature by setting
```
Comment (matrixise) [inline Doc/library/argparse.rst]: If the user would like catch errors manually, the feature can be enable by setting

File: Lib/argparse.py (hunk 2 above)
Comment (matrixise) [inline Lib/argparse.py]: argparser -> ArgumentParser

**Review by matrixise — CHANGES_REQUESTED** [review]
Thank you for your PR, please update with my remarks.

File: Doc/library/argparse.rst (hunk 7 above)
Comment (shihai1991) [reply Doc/library/argparse.rst]: Thanks, Stéphane. done.

**Comment by shihai1991** [comment]
I have made the requested changes; please review again.

File: Doc/library/argparse.rst [hunk 9]
```diff
@@ -179,13 +179,19 @@ ArgumentParser objects
    * allow_abbrev_ - Allows long options to be abbreviated if the
      abbreviation is unambiguous. (default: ``True``)
 
+   * exit_on_error_ - Determines whether or not ArgumentParser exits with
+     error info when an error occur. (default: ``True``)
```
Comment (matrixise) [inline Doc/library/argparse.rst]: when an error occur**s**

**Review by matrixise — CHANGES_REQUESTED** [review]

**Review by matrixise — APPROVED** [review]
Thank you for your PR, I have fixed one Typo, but for the rest, I agree.

File: Doc/library/argparse.rst (hunk 9 above)
Comment (shihai1991) [reply Doc/library/argparse.rst]: ops, thank you :)

**Pull request merged**